review

msharafjr opened the pull request.

msharafjr committed:
add index.js main file

msharafjr committed:
build a skeleton for edigame API

create package.json and empty basic files

msharafjr committed:
add schemas

msharafjr committed:
add packages & add error handling middleware

msharafjr committed:
fill mongod.conf file

msharafjr committed:
test get and post user

msharafjr committed:
add body-parser package & connect mongoose to API

msharafjr committed:
add udpate user and remove user

msharafjr committed:
complete CRUD user

msharafjr committed:
add team schema

msharafjr committed:
add CRUD team

msharafjr committed:
add auth.js file

msharafjr committed:
validate new user

msharafjr committed:
add regex, secretKey & db url to config file, add jwt & bcrypt packages

msharafjr committed:
modify team and user schemas

define comparePassword method and pre hooks to hash user password

msharafjr committed:
add routes

add CRUD teams and admin required

msharafjr committed:
get ready users and toggling ready status

msharafjr committed:
add admin required and leader required

msharafjr committed:
add username, search by user name and reject duplication when update by send response

msharafjr committed:
modify toggle user's ready status

msharafjr committed:
add token required

msharafjr committed:
add get team by nam and get all users of team

msharafjr committed:
add othe teams routes

msharafjr committed:
validate createTeam and add new team to admin schema, add addUserToTeam by admin

msharafjr committed:
add removeUserFromTeam and  modify errorHandler of addUserToTeam

msharafjr committed:
add remove/user and add/user routes

shadyvb commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+node_modules/
+models/mongodb/logs.mongod.log
+models/mongodb/data/db/
+models/mongodb/logs/
```
This location isn't right, `logs` shouldn't exist under `models`. Change the config to logs are outside of this dirctory.

shadyvb commented on controllers/auth.js:
```diff
@@ -0,0 +1,86 @@
+const jwt = require( 'jwt-simple' );
```
Add a header to each file to say what the file is doing, ie file responsibility.

msharafjr replied:
You mean like comment?
// something ?!

shadyvb replied:
Yes, a file header comment. Google it

shadyvb commented on controllers/auth.js:
```diff
@@ -0,0 +1,86 @@
+
+exports.tokenRequired = ( req, res, next ) => {
+    let token = req.headers[ 'x-access-token' ];
+    if ( !token ) return res.status( 403 ).send( 'Token required' );
```
Does this send a JSON object ? It should if not.

shadyvb replied:
All other `send` methods as well.

shadyvb commented on controllers/auth.js:
```diff
@@ -0,0 +1,86 @@
+const config = require( '../models/config' );
+
+exports.loginUser = ( req, res, next ) => {
+    if ( typeof req.body.email !== 'string' )
```
It can still be `string` and be empty. Do a stricter check by checking length of the input, and validate that's an actual email. Same for `password`.

shadyvb commented on controllers/teams.js:
```diff
@@ -0,0 +1,201 @@
+const User = require( '../models/schemas/user' );
+
+exports.getTeams = ( req, res, next ) => {
+    Team.find( {}, ( err, teams ) => {
```
Expect filters here, ie a user shouldn't see teams he's not assigned to unless he's an admin.

msharafjr replied:
I'll put `adminRequired` middleware before it, and make new route for `userTeams`, Sounds good?

shadyvb replied:
Yes, that's better.

shadyvb commented on controllers/teams.js:
```diff
@@ -0,0 +1,201 @@
+}
+
+exports.createTeam = ( req, res, next ) => {
+    if ( typeof req.body.teamName !== 'string' )
```
Same strict checking needed as user/pass note above.

shadyvb commented on controllers/teams.js:
```diff
@@ -0,0 +1,201 @@
+        teamData.admin = req.body.adminId
+
+    let newTeam = new Team( teamData );
+    newTeam.save( ( err, team ) => {
```
Doesn't mongoose support promises ? If yes, use them. Everywhere.

msharafjr replied:
done

shadyvb commented on controllers/teams.js:
```diff
@@ -0,0 +1,201 @@
+            playerId: req.body.playerId
+        });
+
+        team.markModified( 'players');
```
Why do you need to do this ?

msharafjr replied:
I read it from the docs, but turns out it's not required before `save`.

shadyvb replied:
Exactly.

shadyvb commented on controllers/teams.js:
```diff
@@ -0,0 +1,201 @@
+        user.save();
+
+    }).then( () => {
+        return res.sendStatus( 200 );
```
Shouldn't this be `201` ?

msharafjr replied:
for both add/remove user, or `201` for add, `200` for remove?

shadyvb replied:
Correct, 201 is for adding any resource. 200 means action succeeded, so it is a generic one.

shadyvb commented on controllers/teams.js:
```diff
@@ -0,0 +1,201 @@
+        team.markModified( 'players' );
+        team.save();
+
+        user.team = undefined;
```
Users shouldn't have references to teams, it should be the other way around. Because a user might be in multiple team. When you need to get user team, you query the team model not the user model.

ie: its one2many relationship.

msharafjr replied:
OKAY

shadyvb commented on controllers/users.js:
```diff
@@ -0,0 +1,97 @@
+
+	let userData = {};
+
+	if ( req.body.firstName && typeof req.body.firstName === 'string' )
```
This is too much redundant code, create an array of required parameters and loop over them using the same code. ie:

```
let requiredParams = [ 'firstName', 'lastName' , ... ],
error = false;

requiredParams.forEach( ( param ) => {
  if ( error ) return;
  // TODO: add proper form validation somehow
  if ( typeof req.body.username !== 'string' || req.body.username )
    error = 'Missing required' + param;
} );

if ( error ) return res.status( 400 ).send( error );
```

msharafjr replied:
done

shadyvb commented on models/config.js:
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+	port: 3000,
```
What's this ?

msharafjr replied:
port of the server

shadyvb replied:
Why is it here, is the meaning of my question

msharafjr replied:
Hard coding information, so i put any hard coding info in config.js.

shadyvb replied:
An app port doesn't belong in a config file within `models`, that's what I'm saying. Not the correct place for it.

shadyvb reviewed (commented):
Nice work! 

Got some feedback.

msharafjr committed:
commit changes of review from PR#3

using promises instead callbacks for CRUD, add stricter check to validate user inputs, add helpers file and change user schema

msharaf10 commented:
@shadyvb check the changes.
